⭐️ 0.0.1

superturboryan opened the pull request:
See README for package description and usage examples

superturboryan committed:
🚀 Package: add Control + Controllers libraries

- Set minimum iOS version to 15

superturboryan committed:
📝 Package: Add README

superturboryan committed:
🧑‍⚖️ Package: Add MIT license

erikbasargin commented on Sources/Control/Extensions/MPVolumeView.swift:
```diff
@@ -0,0 +1,47 @@
+import MediaPlayer
+
+extension MPVolumeView {
+    
```
What is your opinion on having `Logger` instead of prints? 🐼

Based on current structure, I assume, you can define different Loggers for different areas. 

```suggestion
    private static let log = Logger(subsystem: "com.ControlKit.Control", category: "MPVolumeView_Extensions")
    
```

superturboryan replied:
Great suggestion, I'll add `OSLog` 👌

erikbasargin commented on Sources/Control/Extensions/MPVolumeView.swift:
```diff
@@ -0,0 +1,47 @@
+    
+    static func increaseVolume(_ amount: Float) {
+        guard volume < 1 else {
+            print("🚨 Volume is already at max")
```
```suggestion
            Self.log.debug("🚨 Volume is already at max")
```

erikbasargin commented on Sources/Control/Extensions/MPVolumeView.swift:
```diff
@@ -0,0 +1,47 @@
+    
+    static func decreaseVolume(_ amount: Float) {
+        guard volume > 0 else {
+            print("🚨 Volume is already at min")
```
```suggestion
            Self.log.debug("🚨 Volume is already at min")
```

erikbasargin commented on Sources/Control/Playback.swift:
```diff
@@ -0,0 +1,63 @@
+        ///
+        /// Playback that has been paused by this function can normally be resumed if the app playing the content has not been terminated.
+        public static func togglePlayPause() {
+            try? avAudioSession.setActive(
+                isAudioPlaying,
+                options: .notifyOthersOnDeactivation
+            )
```
Would `AVAudioSession` log an error/warning if `setActive` throws? 👀

What are your thoughts on the following given Playback has a logger?

`private static let log = Logger(subsystem: "com.ControlKit.Control", category: "Playback")`

```suggestion
            do {
                try avAudioSession.setActive(
                    isAudioPlaying,
                    options: .notifyOthersOnDeactivation
                )
            } catch {
                Self.log.warning(error.localizedDescription)
            }
```

superturboryan replied:
Yes, good call, might as well! 🕺

erikbasargin commented on Sources/Control/Volume.swift:
```diff
@@ -0,0 +1,41 @@
+        }
+        
+        public static func decreaseVolume(_ amount: Float = 0.1) {
+            volume -= amount
```
Please correct me if I'm wrong 🐼, seems like if `isMuted` set to `true`, any changes to `Volume.volume` would unmute source, but `isMuted` becomes out of sync 👀

superturboryan replied:
# I totally missed this 🫨

superturboryan replied:
Addressed @erikbasargin 🙋🏻‍♂️

erikbasargin commented on Sources/Control/Volume.swift:
```diff
@@ -0,0 +1,41 @@
+        }
+        
+        /// Subscribe to `isMuted` via ``VolumeController.isMuted`` `@Published` property.
+        static var isMuted = false
```
I guess, `didSet` or `private(set)` is needed here 👀

superturboryan replied:
Using `didSet` instead of the extra `toggleMute()` is much more elegant 💅

Also updated in the `fixup` commit 🆙

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,119 @@
+                clientID: SpotifyConfig.clientID,
+                redirectURL: URL(string: SpotifyConfig.redirectURL)!
+            ),
+            logLevel: .debug
```
What is your opinion on switching the severity of log messages to `error` when there is no DEBUG flag? 👨‍🔧

superturboryan replied:
Sounds good, I'm just wondering how to access the `DEBUG` flag here 🤔

Would you suggest we declare a var somewhere and check that when setting the `logLevel` parameter?

```swift
var isDebug = false
#if DEBUG
isDebug = true
#endif
```

erikbasargin replied:
If there’s no other place where it’s needed, I suppose the local parameter is the best choice. 👍

superturboryan replied:
I've ended up adding the log level as a property of `SpotifyConfig` with the default level as `info`. This way the user of the package can set it themselves 👷‍♂️

superturboryan replied:
I've ended up adding the log level as a property of `SpotifyConfig` with the default level as `info`. This way the user of the package can set it themselves 👷‍♂️

erikbasargin commented on Sources/Control/Extensions/MPVolumeView.swift:
```diff
@@ -0,0 +1,47 @@
+            print("🚨 Volume is already at min")
+            return
+        }
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
```
Out of curiosity 🤓,  why do we need a small delay here?

superturboryan replied:
This _may_ have been copy-pasta'd from an LLM 🤖🍝

Monkey see 🙈 monkey do 🙊

Also updated in the `fixup` commit 🆙

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,119 @@
+    }()
+    
+    @AppStorage("SpotifyAccessToken")
+    private var accessToken: String?
```
Something telling me that it's better to be Keychain 😄

superturboryan replied:
My Security Journey belt should be revoked 🫨

# 🚓👮‍♂️👮‍♀️
![giphy](https://github.com/user-attachments/assets/a9b23feb-1114-441b-bbe0-2e20c0d0f6e8)

superturboryan replied:
Added a property wrapper for `@Control.Keychain` 🔐  

Let me know what you think @erikbasargin

superturboryan replied:
@erikbasargin  as discussed, we'll allow a "data access object" to be injected into `SpotifyController` for the purposes of saving the access token, and provide a default "UserDefaults DAO" in the init. This should provide the best compromise between security and ease-of-use while debugging  🤝

This default parameter should be documented as _unsafe_ in _production_ code, as sensitive data should ~not~ be stored in the keychain! 👮

erikbasargin commented on Sources/Controllers/SystemController.swift:
```diff
@@ -0,0 +1,79 @@
+import Control
+import SwiftUI
+
+/// Controller of controllers 👑
```
👑

superturboryan committed:
🤖 CI: Add lint workflow

- Lints code using SwiftLint

superturboryan committed:
🤖 CI: Add build-package workflow

- Decrement `swift-tools-version` from 6.0 to 5.10.0
- Can be set to 6.0 again when swift-actions/setup-swift supports Swift 6
    - See: https://github.com/swift-actions/setup-swift/pull/684

erikbasargin commented on Sources/Control/Volume.swift:
```diff
@@ -0,0 +1,85 @@
+        ) {
+            if isMuted {
+                isMuted.toggle()
+            }
```
I think this might be redundant since the logic in `volume/set` is the same. 🐼 (the same for `decreaseVolume`)

erikbasargin commented on Package.swift:
```diff
@@ -1,18 +1,32 @@
-// swift-tools-version: 6.0
```
Why not? 😺

superturboryan replied:
I've added a CI workflow to build the package librairies and it's not supported yet 🥲

Let's keep an eye on [this open PR adding support for Swift 6](https://github.com/swift-actions/setup-swift/pull/684).

erikbasargin commented on Sources/Control/Extensions/MPVolumeView.swift:
```diff
@@ -0,0 +1,46 @@
+            log.warning("Volume is already at max")
+            return
+        }
+        volume = max(1, volume + amount)
```
It appears that the parameters used in the functions `increaseVolume` and `decreaseVolume` have been reversed.

`increaseVolume` should utilize the `min`, while `decreaseVolume` should utilize the `max`.

superturboryan replied:
What a catch! ⚾️

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,182 @@
+    
+    public init(
+        config: SpotifyConfig,
+        accessTokenDAO: any DAO<String> = UserDefaultsDAO("SpotifyAccessToken"),
```
It may be preferable to remove the `UserDefaultsDAO(“SpotifyAccessToken”)` argument from the default arguments, as the user may not be aware of this section of the description. While the `accessTokenDAO` parameter is required, it raises a question and necessitates reading the documentation. 🐼

```suggestion
        accessTokenDAO: any DAO<String>,
```

superturboryan replied:
Thanks for calling this out! It certainly raised a lot more questions for me about how the API should be designed...

In the end, I think I'd like to make the `accessTokenDAO` property optional and have the default parameter in the initialiser be `nil`. This way the access token won't be persisted across app launches and the user has to explicitly provide their own DAO. I've updated the readme and doc comments to reflect this change.

This might be the best compromise between ease-of-use and security, wdyt? 👷

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,182 @@
+///
+/// `SpotifyController` facilitates the connection to and interaction with the Spotify app, enabling playback control and handling the authorization flow.
+///
+/// - Parameters:
+///   - config: The `SpotifyConfig` object used to configure the controller.
+///   If `config` is `.empty`, the controller will not attempt to connect automatically.
+///   The default log level is ``SPTAppRemoteLogLevel//debug``.
+///   - accessTokenDAO: An object conforming to `DAO<String>` that manages the storage and retrieval of the Spotify access token. By default, this uses a `UserDefaultsDAO` with a key of `"SpotifyAccessToken"`.
+///   - autoConnect: A Boolean value that determines whether the controller should attempt to connect to Spotify automatically upon initialization.
+///   The default value is `true`. If `true` and `config` is not `.empty`, the controller will initiate the connection.
+///
+/// - Warning: Saving the Spotify access token using **``UserDefaults``** is not recommended for production apps 🚨
+///   Prefer a `DAO` that persists values to the Keychain 🔐
```
I guess it can be added as a description for the init method with the note that `UserDefaultsDAO("SpotifyAccessToken")` can be used for testing.

superturboryan replied:
Sounds good, I've split this documentation across the struct declaration + init method 🤝

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,182 @@
+///
+public final class SpotifyController: NSObject, ObservableObject {
+    
+    @Published public var isPlaying: Bool = false
```
Should it be modifiable from outside? 👀

```suggestion
    @Published public private(set) var isPlaying: Bool = false
```

superturboryan replied:
**Certainly not** 🙅‍♂️

superturboryan commented on .swiftlint.yml:
```diff
@@ -0,0 +1,74 @@
+  - void_function_in_ternary
+#  - force_cast
+
+opt_in_rules:
```
A lot of the rules here are not necessary for this package - I've copy-pasta'd them from another project where I had a _stricter_ linting setup 🧑‍⚖️

erikbasargin commented on Sources/Control/Volume.swift:
```diff
@@ -0,0 +1,79 @@
+                MPVolumeView.volume
+            }
+            set {
+                if newValue != 0 && isMuted {
+                    isMuted.toggle()
+                }
```
Would you like to maintain the `isMuted` property set to `true` when the `volume` is explicitly set to 0?

superturboryan replied:
I'm trying to wrap my head around this one and I think the code matches what you describe? 🤔

The idea is to "unmute if not zero"

erikbasargin commented on Sources/Controllers/AppleMusicController.swift:
```diff
@@ -0,0 +1,35 @@
+/// Wrapper for ``Control/Playback/AppleMusic`` - use this to control playback from the Apple Music app and subscribe to its state.
+public final class AppleMusicController: ObservableObject, PlaybackController {
+    
+    @Published public var isPlaying: Bool = Control.Playback.AppleMusic.isPlaying
```
Should it be modifiable from outside? 👀

```suggestion
    @Published public private(set) var isPlaying: Bool = Control.Playback.AppleMusic.isPlaying
```

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,202 @@
+            ),
+            logLevel: config.logLevel
+        )
+        remote.connectionParameters.accessToken = try? accessTokenDAO?.get() // Non-blocking failure
```
[Minor] What are your thoughts on adding logging here? 🙂

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,202 @@
+            Self.log.error("Failed to parse access token from URL: \(url)")
+            return
+        }
+        try? accessTokenDAO?.save(newToken) // Non-blocking failure
```
[Minor] What are your thoughts on adding logging here? 🙂 (the same for `disconnect` function)

erikbasargin commented on Sources/Controllers/SpotifyController.swift:
```diff
@@ -0,0 +1,202 @@
+            let parameters = remote.authorizationParameters(from: url),
+            let newToken = parameters[SPTAppRemoteAccessTokenKey]
+        else {
+            Self.log.error("Failed to parse access token from URL: \(url)")
```
I suppose it would be more prudent to implement a privacy level on locations where sensitive data could potentially be logged. 🤓

```suggestion
            Self.log.error("Failed to parse access token from URL: \(url, privacy: .private)")
```

erikbasargin commented on Sources/Controllers/SystemController.swift:
```diff
@@ -0,0 +1,98 @@
+    
+    @Published public var appleMusic = AppleMusicController()
+    @Published public var spotify: SpotifyController!
+    @Published public var volume = VolumeController()
```
Should all these published parameters be modifiable from outside? 👀

erikbasargin commented on Sources/Control/Playback.swift:
```diff
@@ -0,0 +1,73 @@
+
+public extension Control.Playback {
+    
+    enum AppleMusic {
```
Seems like it can be marked with MainActor. What is your opinion? 🐼

```suggestion
    @MainActor
    enum AppleMusic {
```

erikbasargin commented on Sources/Controllers/VolumeController.swift:
```diff
@@ -0,0 +1,43 @@
+@MainActor
+public final class VolumeController: ObservableObject {
+    
+    @Published public var volume: Float = Control.Volume.volume
```
It appears that this property can also be marked as `private(set)` since there is no subscription to changes from outside sources? 👀
